onnxruntime/core/providers/cuda/igemm.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "igemm.h"
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/shared_inc/cuda_call.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+#if CUDA_VERSION >= 10010
+void LtIgemmTensor(int m,
+                   int n,
+                   int k,
+                   int32_t alpha_matmul,
+                   int32_t beta_matmul,
+                   const int8_t* a,
+                   int lda,
+                   const int8_t* b,
+                   int ldb,
+                   int32_t* c,
+                   int ldc,
+                   const CudaKernel* cuda_kernel,
+                   cublasLtHandle_t lt_handle) {
+  // Create descriptors for the original matrices
+  cublasLtMatrixLayout_t a_desc = NULL;
+  cublasLtMatrixLayout_t b_desc = NULL;
+  cublasLtMatrixLayout_t c_desc = NULL;
+  CUBLAS_CALL_THROW(cublasLtMatrixLayoutCreate(&a_desc, CUDA_R_8I, m, k, lda));
+  CUBLAS_CALL_THROW(cublasLtMatrixLayoutCreate(&b_desc, CUDA_R_8I, n, k, ldb));
+  CUBLAS_CALL_THROW(cublasLtMatrixLayoutCreate(&c_desc, CUDA_R_32I, m, n, ldc));
+
+  // Set A and C row major order.
+  // No need for B because B need to be transposed
+  cublasLtOrder_t row_order = CUBLASLT_ORDER_ROW;
+  CUBLAS_CALL_THROW(cublasLtMatrixLayoutSetAttribute(a_desc, CUBLASLT_MATRIX_LAYOUT_ORDER, &row_order, sizeof(row_order)));
+  CUBLAS_CALL_THROW(cublasLtMatrixLayoutSetAttribute(c_desc, CUBLASLT_MATRIX_LAYOUT_ORDER, &row_order, sizeof(row_order)));
+
+  int lda_transform = 32 * m;
+  int ldb_transform = 32 * roundoff(n, 8);
```

Comment:
why ldb need roundoff, but a and b need not?

Reply:
A and C use layout CUBLASLT_ORDER_COL32, but B uses layout CUBLASLT_ORDER_COL4_4R2_8C. Will add more comment

---
In reply to: [404461348](https://github.com/microsoft/onnxruntime/pull/3413#discussion_r404461348) [](ancestors = 404461348)